projects/concurrency/README.md

```diff
@@ -0,0 +1,308 @@
+<!--forhugo
++++
+title="Concurrency"
++++
+forhugo-->
+
+In this project you're going to learn about how to safely use multiple threads. You will learn about concurrency, thread safety, and synchronisation.
+
+Timebox: 3 days
+
+## Learning Objectives:
+
+- Suggest orders of magnitude for the speed of different read operations
+- Define the term "memory barrier"/"fence" in the context of memory models
+- Explain when and why synchronisation is needed
+- Choose appropriate synchronisation primitives for specific problems from atomics, wait-groups, mutexes/locks, read-write locks, and channels
+- Identify the critical section across multiple statements
+
+## Background
+
+### Memory models
+
+As you will have learnt in Chapter 7 of [How Computers Really Work](https://www.howcomputersreallywork.com/), some memory is cheap but slow, and other memory is fast but expensive. CPU manufacturers add fast per-core caches in front of cheap system-wide memory. This causes problems when multiple cores try to share the same item in memory. Each core's cache may get its own copy of that memory, and if one core changes the contents of that data, the other core won't see that change.
+
+You can see estimated speeds of different memory (and other) read operations in ["Latency Numbers Every Programmer Should Know"](https://gist.github.com/jboner/2841832). Read this and make sure you undersatnd it. This suggests that reading from shared main memory is 100-1000x slower than reading from a per-core L1 cache.
+
+This set of projects will introduce different ways we can handle this problem.
+
+Read [the concurrency section](https://golangbot.com/learn-golang-series/#concurrency) of [golangbot.com](https://golangbot.com/) (i.e. chapters 20-25). Run the examples and make sure you understand their output.
+
+## Projects
+
+### Single-variable concurrency
+
+Let's look at the example buggy code from the Mutexes chapter of golangbot:
+
+```go
+package main
+
+import (
+	"fmt"
+	"sync"
+)
+
+var x = 0
+
+func increment(wg *sync.WaitGroup) {
+	x = x + 1
+	wg.Done()
+}
+
+func main() {
+	var w sync.WaitGroup
+	for i := 0; i < 1000; i++ {
+		w.Add(1)
+		go increment(&w)
+	}
+	w.Wait()
+	fmt.Println("final value of x", x)
+}
+```
+
+In an ideal world, this program would always output 1000.
+
+In reality, if you run it 5 times, you will probably get 5 different answers.
```

Comment:
should we suggest here that they actually do this? i like to be practical!
in fact, we could modify this code to move the contents of main off to another function and run that N times, so they can see for themselves that they get different answers all the time.